### ISSUE TITLE
Changing remote_user in hybrid setups problematic

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#21866]
##### ISSUE TYPE

 - Docs Report

##### COMPONENT NAME

core

##### ANSIBLE VERSION

v2.3 and older

##### OS / ENVIRONMENT

Hybrid environment with Windows and Linux

##### SUMMARY

In a hybrid environment, the default `ansible_user` points to either `root` or `Administrator`.

Which means we have to set the default `ansible_user` by using `group_vars`. So I have `group_vars/windows` defining `ansible_user: Administrator`.

But because this is done, using `remote_user` on the task-level, no longer works. The `group_vars` `ansible_user` wins from the task-level `remote_user`. And even wins from the task-level task_vars `ansible_user`.

In other words, there is no possibility to have a task use a different user.

**Comment by dagwieers** [comment]
cc: @nitzmahone @jctanner @jimi-c 

References: #4688 #7673

**Comment by bcoca** [comment]
task level ansible_user should win over inventory defined ansible_user

**Comment by dagwieers** [comment]
@bcoca But does it win over group_vars/windows ansible_user ?

**Comment by nitzmahone** [comment]
dagwieers: this is specifically setting vars: ansible_user on the task, not remote_user

**Comment by bcoca** [comment]
@dagwieers it should by variable precedence and a quick test show it does, if it does not, that is a bug

**Comment by dagwieers** [comment]
What I think is confusing here, is the `remote_user` at the task-level does not seem to be listed in the variable precedence.
```yaml
  - win_shell: setup.exe ...
    remote_user: '{{ domain }}\{{ windows_scvmm_user }}'
```
You would expect that it either is a task variable, or a core directive (which is processed locally). It does not fit in any other category. In any case, inventory group_vars are completely at the end of the variable precedence order, so you'd never expect that one to win.

Why else would `remote_user` be a special task-level directive ?

So this pretty much feels like a bug.

**Comment by bcoca** [comment]
inventory:
```
test1
   ansbile_user: lola
```
play

```
- hosts: test1
  gather_facts: False
  tasks:
   - ping:
     vars:
        ansible_user: notlola
```
-vvv output:
```
Using module file /home/bcoca/work/ansible/lib/ansible/modules/system/ping.py
<192.168.0.4> ESTABLISH SSH CONNECTION FOR USER: notlola

```

**Comment by bcoca** [comment]
remote_user is NOT a variable, its a directive, we do document that 'connection variables' will override directives as we consider host connection variables MORE specific than ansible.cfg, command line switches and play directives.

**Comment by dagwieers** [comment]
The documentation does not list "connection variables" as part for the variable precedence (it does for 1.x, but doesn't for 2.0). However, ansible_user is also a connection variable (and specifically mentioned as such for the 1.x precedence list).

So it seems to be contradicting, both the 1.x as the 2.0 documentation seems to be wrong in our case. Should we update the documentation ?

**Comment by bcoca** [comment]
it exists this way so you can change the 'login user' in the general case, but allow special cases in your network to use a specific user always. 

For example, your DBA team has a requirement that only the 'common_dba' user is used on their boxes. You normally logon as yourself to most boxes, but now you have a play that uses the 'webuser', but the DBA machines still have the 'common_dba' user required. So you would use `remote_user=webuser` in the play  and `ansible_user=common_dba` in the db machine inventory group.

This is not your scenario, but we needed to design Ansible to handle most if not all possible scenarios.

**Comment by bcoca** [comment]
@dagwieers ansible_user follows normal variable precedence for itself, 'connection variables' in general override directives, which are NOT variables.

**Comment by bcoca** [comment]
from docs "Another important thing to consider (for all versions) is that connection specific variables override config, command line and play specific options and directives. For example:"

**Comment by dagwieers** [comment]
bcoca: For remote_user in the play, I could agree, but a remote_user at the task-level ?

PS Ah ok, I searched specifically for "connection variable", not for "connection specific variable". But I do think it should be in the variable precedence list specifically.
PS2 It states play-specific options and directives, not task-specific options and directives.

**Comment by bcoca** [comment]
refers to everything in play, but i would be glad to clarify that.

and yes, even at task level you might want to override for those hosts, in my example above it can be a single task that needs webuser/common_dba user.

You CAN override both ways in current system, its complex but allows for complex setups (which sadly exist commonly). We could make it simpler, but then we wouldn't work for those cases.

**Comment by dagwieers** [comment]
@bcoca I don't mind using task-vars for this, but this *is* confusing and contra-intuitive (i.e. exceptional). The documentation could clarify it better (and specifically adding the variables/directives like remote_user would be a plus, so people could find it when searching).

**Comment by bcoca** [comment]
^ bikeshed away, lets get some docs that clarify this

**Comment by bcoca** [comment]
im not saying its not confusing, but sadly the problem addressed is complex and I don't see a simple way to allow for all scenarios and still have an intuitive and simple interface

**Comment by dagwieers** [comment]
I don't know how to explain exactly how this works, but don't mind reviewing the new wording to see if it would have clarified it for me. But I do think "connection variables" or "connection directives" should be mentioned somewhere **in the variable precedence list** if they break with convention.

Updated to clarify that they need to be listed in the precedence list (as they are not listed there).

**Comment by bcoca** [comment]
they are .. not sure where else they need mentioning

**Comment by dagwieers** [comment]
@bcoca Thanks !
